### PR TITLE
[dtensor][debug] Creating recipe allowing users to learn how CommDebugMode works and use visual browser

### PR DESCRIPTION
**Pull request by sinhaanshul**
Fixes #ISSUE_NUMBER

## Description
I created a recipe showing users how to use CommDebugMode to obtain the results they want. In addition, I have embedded an html file so that users can access and use the visual browser here. 

## Checklist

- [ ] The issue that is being fixed is referred in the description (see above "Fixes #ISSUE_NUMBER")
- [ ] Only one issue is addressed in this pull request
- [ ] Labels from the issue that this PR is fixing are added to this pull request
- [ ] No unnecessary issues are included into this pull request.

cc @wconstab @osalpekar @H-Huang @kwen2501

### CODE REVIEW

File: recipes_source/distributed_comm_debug_mode.rst [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+- PyTorch 2.2
+
+
+What is CommDebugMode and why is it useful
+------------------------------------------
+As the size of models continues to increase, users are seeking to leverage various combinations
+of parallel strategies to scale up distributed training. However, the lack of interoperability
+between existing solutions poses a significant challenge, primarily due to the absence of a
+unified abstraction that can bridge these different parallelism strategies. To address this
+issue, PyTorch has proposed DistributedTensor(DTensor) which abstracts away the complexities of
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:19]: link to https://pytorch.org/docs/stable/distributed.tensor.parallel.html ?

File: recipes_source/distributed_comm_debug_mode.rst [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+Prerequisites:
+
+- `Distributed Communication Package - torch.distributed <https://pytorch.org/docs/stable/distributed.html>`__
+- Python 3.8 - 3.11
+- PyTorch 2.2
+
+
+What is CommDebugMode and why is it useful
+------------------------------------------
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:14]:
```suggestion
------------------------------------------

```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+As the size of models continues to increase, users are seeking to leverage various combinations
+of parallel strategies to scale up distributed training. However, the lack of interoperability
+between existing solutions poses a significant challenge, primarily due to the absence of a
+unified abstraction that can bridge these different parallelism strategies. To address this
+issue, PyTorch has proposed DistributedTensor(DTensor) which abstracts away the complexities of
+tensor communication in distributed training, providing a seamless user experience. However,
+this abstraction creates a lack of transparency that can make it challenging for users to
+identify and resolve issues. To address this challenge, CommDebugMode, a Python context manager
+will serve as one of the primary debugging tools for DTensors, enabling users to view when and
+why collective operations are happening when using DTensors, addressing this problem.
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:24]:
```suggestion
why collective operations are happening when using DTensors, effectively addressing this issue.
```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+
+How to use CommDebugMode
+------------------------
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:28]:
```suggestion
------------------------

```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+Using CommDebugMode and getting its output is very simple.
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:29]:
```suggestion
Here is how you can use ``CommDebugMode``:
```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+              FORWARD PASS
+                *c10d_functional.all_reduce: 1
+                MLPModule.net1
+                MLPModule.relu
+                MLPModule.net2
+                  FORWARD PASS
+                    *c10d_functional.all_reduce: 1
+    """
+
+All users have to do is wrap the code running the model in CommDebugMode and call the API that
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:65]:
```suggestion
To use ``CommDebugMode`, you must wrap the code running the model in ``CommDebugMode`` and call the API that
```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+they want to use to display the data. One important thing to note is that the users can use a noise_level
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:66]:
```suggestion
you want to use to display the data. You can also use a ``noise_level``
```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+arguement to control how much information is displayed to the user. The information below shows what each
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:67]:
```suggestion
argument to control the verbosity level of displayed information. Here is what each
```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+noise level displays
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:68]:
```suggestion
noise level displays:
```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+
+| 0. prints module-level collective counts
+| 1. prints dTensor operations not included in trivial operations, module information
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:71]:
```suggestion
| 1. Prints ``dTensor`` operations not included in trivial operations, module information
```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+                  FORWARD PASS
+                    *c10d_functional.all_reduce: 1
+    """
+
+All users have to do is wrap the code running the model in CommDebugMode and call the API that
+they want to use to display the data. One important thing to note is that the users can use a noise_level
+arguement to control how much information is displayed to the user. The information below shows what each
+noise level displays
+
+| 0. prints module-level collective counts
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:70]:
```suggestion
| 0. Prints module-level collective counts
```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+| 1. prints dTensor operations not included in trivial operations, module information
+| 2. prints operations not included in trivial operations
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:72]:
```suggestion
| 2. Prints the operations that are not included in trivial operations
```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+| 3. prints all operations
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:73]:
```suggestion
| 3. Prints all operations
```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+
+In the example above, users can see in the first picture that the collective operation, all_reduce, occurs
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:75]: Should there be an image?
```suggestion
In the example above, you can see in the first picture that the collective operation, ``all_reduce``, occurs
```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+once in the forward pass of the MLPModule. The second picture provides a greater level of detail, allowing
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:76]: There should be an image again?
```suggestion
once in the forward pass of the ``MLPModule``. The second picture provides a greater level of detail, allowing
```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+users to pinpoint that the all-reduce operation happens in the second linear layer of the MLPModule.
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:77]:
```suggestion
you to pinpoint that the all-reduce operation happens in the second linear layer of the ``MLPModule``.
```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+            <input type="file" id="file-input" accept=".json">
+          </div>
+        <div id="tree-container"></div>
+        <script src="https://cdn.jsdelivr.net/gh/pytorch/pytorch@main/torch/distributed/_tensor/debug/comm_mode_broswer_visual.js"></script>
+    </body>
+    </html>
+
+Conclusion
+------------------------------------------
+In conclusion, we have learned how to use CommDebugMode in order to debug Distributed Tensors
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:193]:
```suggestion

In this recipe, we have learned how to use ``CommDebugMode`` to debug Distributed Tensors.
```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+and can use future json dumps in the embedded visual browser.
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:194]:
```suggestion
You can use your own JSON outputs in the embedded visual browser.
```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+
+For more detailed information about CommDebugMode, please see
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:196]:
```suggestion
For more detailed information about ``CommDebugMode``, see
```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 20]
```diff
@@ -0,0 +1,197 @@
...
+https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/examples/comm_mode_features_example.py
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:197]:
```suggestion
`comm_mode_features_example.py
< https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/examples/comm_mode_features_example.py>`__
```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 21]
```diff
@@ -0,0 +1,197 @@
+Using CommDebugMode
+=====================================================
+
+**Author**: `Anshul Sinha <https://github.com/sinhaanshul>`__
+
+Prerequisites:
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:6]:
```suggestion
Prerequisites
------------------
```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 22]
```diff
@@ -0,0 +1,197 @@
...
+
+- `Distributed Communication Package - torch.distributed <https://pytorch.org/docs/stable/distributed.html>`__
+- Python 3.8 - 3.11
+- PyTorch 2.2
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:10]:
```suggestion
- PyTorch 2.2 or later
```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 23]
```diff
@@ -0,0 +1,197 @@
+Using CommDebugMode
+=====================================================
+
+**Author**: `Anshul Sinha <https://github.com/sinhaanshul>`__
+
+Prerequisites:
+
+- `Distributed Communication Package - torch.distributed <https://pytorch.org/docs/stable/distributed.html>`__
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:8]: I think it's included with PyTorch, so we probably don't need to state this explicitly?

File: recipes_source/distributed_comm_debug_mode.rst [hunk 24]
```diff
@@ -0,0 +1,197 @@
...
+            *c10d_functional.all_reduce: 1
+            MLPModule
+              FORWARD PASS
+                *c10d_functional.all_reduce: 1
+                MLPModule.net1
+                MLPModule.relu
+                MLPModule.net2
+                  FORWARD PASS
+                    *c10d_functional.all_reduce: 1
+    """
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:63]: This sounds like a separate output to me not the part, you want users to copy paste?
```suggestion

 This is what the output looks like for a MLPModule at noise level 0:
 
 .. code-block:: bash
    Expected Output:
        Global
          FORWARD PASS
            *c10d_functional.all_reduce: 1
            MLPModule
              FORWARD PASS
                *c10d_functional.all_reduce: 1
                MLPModule.net1
                MLPModule.relu
                MLPModule.net2
                  FORWARD PASS
                    *c10d_functional.all_reduce: 1
```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 25]
```diff
@@ -0,0 +1,202 @@
...
+you want to use to display the data. You can also use a ``noise_level`` argument to control the verbosity
+level of displayed information. Here is what each noise level displays:
+
+| 0. Prints module-level collective counts
+| 1. Prints dTensor operations not included in trivial operations, module information
+| 2. Prints operations not included in trivial operations
+| 3. Prints all operations
+
+In the example above, you can see that the collective operation, all_reduce, occurs once in the forward pass
+of the ``MLPModule``. Furthermore, you can use ``CommDebugMode`` pinpoint that the all-reduce operation happens
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:79]:
```suggestion
of the ``MLPModule``. Furthermore, you can use ``CommDebugMode`` to pinpoint that the all-reduce operation happens
```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 26]
```diff
@@ -0,0 +1,202 @@
...
+in the second linear layer of the ``MLPModule``.
+
+
+Below is the interactive module tree visualization that users can upload their JSON dump to:
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:83]:
```suggestion
Below is the interactive module tree visualization that you can use to upload your own JSON dump:
```

**Review by svekars — COMMENTED** [review]
minor nits

File: recipes_source/distributed_comm_debug_mode.rst [hunk 27]
```diff
@@ -0,0 +1,202 @@
...
+How to use CommDebugMode
+------------------------
+
+Here is how you can use ``CommDebugMode``:
+
+.. code-block:: python
+
+    comm_mode = CommDebugMode()
+        with comm_mode:
+            output = model(inp)
```
Comment (wanchaol) [inline recipes_source/distributed_comm_debug_mode.rst:37]: maybe explain a bit in the comment that `model` here could be a MLP model that applied TP?

File: recipes_source/distributed_comm_debug_mode.rst [hunk 28]
```diff
@@ -0,0 +1,202 @@
...
+What is CommDebugMode and why is it useful
+------------------------------------------
+As the size of models continues to increase, users are seeking to leverage various combinations
+of parallel strategies to scale up distributed training. However, the lack of interoperability
+between existing solutions poses a significant challenge, primarily due to the absence of a
+unified abstraction that can bridge these different parallelism strategies. To address this
+issue, PyTorch has proposed `DistributedTensor(DTensor)
+<https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/examples/comm_mode_features_example.py>`_
+which abstracts away the complexities of tensor communication in distributed training,
+providing a seamless user experience. However, this abstraction creates a lack of transparency
```
Comment (wanchaol) [inline recipes_source/distributed_comm_debug_mode.rst:21]:
```suggestion
providing a seamless user experience. However, when dealing with existing parallelism solutions and developing parallelism solutions using the unified abstraction like DTensor, the lack of transparency about what and when the collective communications happens under the hood could make it challenging for advanced users to identify and resolve issues.
```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 29]
```diff
@@ -0,0 +1,202 @@
...
+          </div>
+        <div id="tree-container"></div>
+        <script src="https://cdn.jsdelivr.net/gh/pytorch/pytorch@main/torch/distributed/_tensor/debug/comm_mode_broswer_visual.js"></script>
+    </body>
+    </html>
+
+Conclusion
+------------------------------------------
+
+In this recipe, we have learned how to use ``CommDebugMode`` to debug Distributed Tensors. You can use your
```
Comment (wanchaol) [inline recipes_source/distributed_comm_debug_mode.rst:197]:
```suggestion
In this recipe, we have learned how to use ``CommDebugMode`` to debug Distributed Tensors and parallelism solutions that uses communication collectives with PyTorch
```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 30]
```diff
@@ -0,0 +1,202 @@
...
+                  FORWARD PASS
+                    *c10d_functional.all_reduce: 1
+    """
+
+To use ``CommDebugMode``, you must wrap the code running the model in ``CommDebugMode`` and call the API that
+you want to use to display the data. You can also use a ``noise_level`` argument to control the verbosity
+level of displayed information. Here is what each noise level displays:
+
+| 0. Prints module-level collective counts
+| 1. Prints dTensor operations not included in trivial operations, module information
```
Comment (wanchaol) [inline recipes_source/distributed_comm_debug_mode.rst:74]:
```suggestion
| 1. Prints DTensor operations (not including trivial operations), module sharding information
```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 31]
```diff
@@ -0,0 +1,202 @@
...
+| 2. Prints operations not included in trivial operations
```
Comment (wanchaol) [inline recipes_source/distributed_comm_debug_mode.rst:75]:
```suggestion
| 2. Prints tensor operations (not including trivial operations)
```

**Review by wanchaol — COMMENTED** [review]
Thanks for adding the recipe! Please see inlined comments

File: recipes_source/distributed_comm_debug_mode.rst [hunk 32]
```diff
@@ -0,0 +1,206 @@
+Using CommDebugMode
+=====================================================
+
+**Author**: `Anshul Sinha <https://github.com/sinhaanshul>`__
+
+Prerequisites
+
+- Python 3.8 - 3.11
+- PyTorch 2.2 or later
+
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:10]:
```suggestion

In this tutorial, we will explore how to use CommDebugMode with PyTorch's DistributedTensor (DTensor) for debugging by tracking collective operations in distributed training environments.

Prerequisites
---------------------
* Python 3.8 - 3.11
* PyTorch 2.2 or later

```

File: recipes_source/distributed_comm_debug_mode.rst [hunk 33]
```diff
@@ -0,0 +1,210 @@
+Using CommDebugMode
+=====================================================
+
+**Author**: `Anshul Sinha <https://github.com/sinhaanshul>`__
+
+
+In this tutorial, we will explore how to use CommDebugMode with PyTorch's DistributedTensor (DTensor) for debugging by tracking collective operations in distributed training environments.
```
Comment (svekars) [inline recipes_source/distributed_comm_debug_mode.rst:7]:
```suggestion
In this tutorial, we will explore how to use CommDebugMode with PyTorch's
DistributedTensor (DTensor) for debugging by tracking collective operations in distributed training environments.
```

**Review by wanchaol — APPROVED** [review]
great work! looks good to me!

**Comment by svekars** [comment]
@sinhaanshul can you please sign the CLA above?

**Pull request closed without merging**